Add the Reaper's Lighter spell

datlo opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

This is not an April Fool's PR, I just happen to be on break today and felt like coding this.

## What Does This PR Do
This PR adds the Reaper's Lighter spell, a mobility spell alternative to the classic jaunt, based on the game Deadbolt whose protagonist has a similar ability.

**Reaper's Lighter :**

With a base cooldown of 12 seconds and a cost of 1 spell point, when cast near a vent or scrubber, you will immediately turn to smoke and enter the vent, even if the vent is welded. You can also leave through welded vents.

The spell's cooldown does not start counting down until you leave vents through any mean.

You are immune to temperature, pressure, and do not need to breath for the duration of this ventcrawl. 

If you are still ventcrawling after 12 seconds, you will be pulled back to your starting vent's turf, so wizards cannot stay forever hidden in vents. There is a warning 6 seconds in to make you aware of this, aswell as an alert with a countdown on the screen (under hunger, oxygen, etc...)

**Other Tweaks :**

The 'Gunslinging Reaper', originally also inspired by Deadbolt when I added loadouts, was tweaked to include this spell. It loses Jaunt and Blink, but gains the Reaper's Lighter, Smoke, Knock, and Greater Forcewall. Its coat now also matches wizard robes resistance using a unique subtype. It now has an extra 2 speedloaders because turns out paradise has a lot of players on your average round, and 4 wasn't cutting it when I got to playtest the loadout without these changes.

To prevent reapers from dropping their gear when using the Smoke spell, and because it makes sense anyway, mobs with NO_BREATH no longer inhales choking smoke and chem gases.

Wizards who are hit by the Mime Malaise spell will now be changed into mime robes instead of regular mime outfit. Mime robes were not in the codebase when I originally made the spell, so changing it now as it should be.

Refactored countdown alerts to be more flexible, was previously strictly for polling ghosts.

## Why It's Good For The Game
<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

This spell provides an interesting alternative mobility option to Ethereal Jaunt, with a lower cooldown and cost but a more restricted method of travel. From testing, the spell feels very intuitive to use (Click spell => Immediately ventcrawl), and I hope will be satisfying both to play with and against.

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif or mp4 of your feature if you want. -->

Fixed to say 'within 12 seconds' as it should already
![image](https://user-images.githubusercontent.com/22121511/168114200-d47ba364-1d3a-4090-b4ef-1cb046a1db58.png)

![image](https://user-images.githubusercontent.com/22121511/168114313-5385f75a-4586-4ae3-8f72-36fdabf2b5c3.png)


![reaper2](https://user-images.githubusercontent.com/22121511/161275754-8b73a240-4aad-4474-8ef6-1338fa61bcfb.png)

![image](https://user-images.githubusercontent.com/22121511/168114070-5d8f0767-3a41-4cd9-83fb-2b8a3c0e5224.png)

## Changelog
:cl:
add: Added the Reaper's Lighter spell, allowing wizards to turn into smoke and ventcrawl for a limited period of time.
tweak: Replaced Ethereal Jaunt and Blink in the Gunslinging Reaper wizard loadout by the Reaper's Lighter, Smoke, Knock and Greater Forcewall.
tweak: Added two .357 speedloaders to the Gunslinging Reaper loadout.
tweak: The coat in the gunslinging reaper outfit is now reinforced, matching wizard robes resistances.
tweak: Mobs that do not breath no longer inhales choking smoke and chem gases.
tweak: Wizards hit by the Mime malaise spell are now given mime wizard robes.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

datlo commented:
few tweaks needed still, hold on the review

hal9000PR commented on code/__HELPERS/traits.dm:
```diff
@@ -183,6 +183,7 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_NOEXAMINE			"no_examine"
 #define TRAIT_NOPAIN			"no_pain"
 #define TRAIT_FORCE_DOORS 		"force_doors"
+#define TRAIT_BYPASS_WELDED_VENTS		"force_doors"
```
Copy pasted define?

datlo replied:
Ah fack. Good catch

datlo commented:
Ready for review/merge

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+	smoke.set_up(1, 0, T)
+	smoke.start()
+
+/obj/effect/proc_holder/spell/reaper_lighter/proc/toggle_traits(mob/user)
+	if(!in_progress)
+		ADD_TRAIT(user, TRAIT_RESISTHEAT, "reaper's lighter")
+		ADD_TRAIT(user, TRAIT_RESISTCOLD, "reaper's lighter")
+		ADD_TRAIT(user, TRAIT_RESISTHIGHPRESSURE, "reaper's lighter")
+		ADD_TRAIT(user, TRAIT_RESISTLOWPRESSURE, "reaper's lighter")
+		ADD_TRAIT(user, TRAIT_NOBREATH, "reaper's lighter")
+		ADD_TRAIT(user, TRAIT_BYPASS_WELDED_VENTS, "reaper's lighter")
+	else
+		REMOVE_TRAIT(user, TRAIT_RESISTHEAT, "reaper's lighter")
+		REMOVE_TRAIT(user, TRAIT_RESISTCOLD, "reaper's lighter")
+		REMOVE_TRAIT(user, TRAIT_RESISTHIGHPRESSURE, "reaper's lighter")
+		REMOVE_TRAIT(user, TRAIT_RESISTLOWPRESSURE, "reaper's lighter")
+		REMOVE_TRAIT(user, TRAIT_NOBREATH, "reaper's lighter")
+		REMOVE_TRAIT(user, TRAIT_BYPASS_WELDED_VENTS, "reaper's lighter")
```
please use the `MAGIC_TRAIT` define here.

datlo replied:
This is intentional, using the MAGIC_TRAIT define would overwrite traits added by other spells, while these traits are just temporary and constantly get turned on and off

hal9000PR replied:
understandable. please make a define in `traits.dm` under `unique trait sources` . defines are much preferred to strings in these situations.

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+	if(!target)
+		charge_counter = charge_max
+		return FALSE
+	toggle_traits(usr)
```
```suggestion
	toggle_traits(user)
```

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+			in_progress = FALSE
+			warned = FALSE
+			start_recharge()
+			spawn_smoke(starting_vent_turf)
+			user.forceMove(starting_vent_turf)
+
```
might want to null out the starting vent turf here. I am not sure if turfs GC like everything else does because they are never really deleted, but its still good practice to do so.

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+/obj/effect/proc_holder/spell/reaper_lighter/cast(list/targets, mob/user = usr)
+	var/obj/machinery/atmospherics/unary/target = targets[1]
+
+	if(!target)
+		charge_counter = charge_max
+		return FALSE
```
I am pretty sure, @farie82 would know more, but if a spell fails to find any target it automatically refunds itself.

datlo replied:
It doesn't at this stage of spellcasting, this is a failsafe that shouldnt trigger anyway, there's already a target check earlier in the process, but in some edge cases this can get triggered.

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+	desc = "Use this spell next to a vent to turn to smoke and crawl through it, even if the vents are welded. If you do not leave the vents within 15 seconds, you will be pulled back to where you started."
+
+	school = "transmutation"
+	charge_max = 150
```
```suggestion
	charge_max = 15 SECONDS
```

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+	clothes_req = FALSE
+	invocation = "none"
+	invocation_type = "none"
+	cooldown_min = 30 //3 seconds reduction per rank
```
```suggestion
	cooldown_min = 3 SECONDS //3 seconds reduction per rank
```

datlo commented:
Disclaimer : I forgot to note a small buff to the gunslinging reaper outfit. Its provided coat now matches wizard robes in resistance, so it's really just a style buff if anything. I added the change to the changelog and PR description.

hal9000PR commented on code/datums/spells/reaper_lighter.dm:
```diff
@@ -0,0 +1,85 @@
+	time_to_warn = world.time + ventcrawl_warning_time
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/effect/proc_holder/spell/reaper_lighter/process()
+	var/mob/user = action.owner
+	if(in_progress)
+		if(!warned && world.time > time_to_warn)
+			warned = TRUE
+			to_chat(user, "<span class='warning'>You will be pulled back to your entry vent soon!</span>")
+		if(world.time > time_to_leave)
+			toggle_traits(user)
+			in_progress = FALSE
+			warned = FALSE
+			start_recharge()
+			spawn_smoke(starting_vent_turf)
+			user.forceMove(starting_vent_turf)
+			starting_vent_turf = null
+	else
+		..()
```
I think this logic would be better being part of a `status_effect`. rather than making use of the spells process. You could have the traits be applied when you add the status_effect and removed when it runs out. then you wouldn't need the `toggle_traits` proc. and you wouldn't need the `smoke_crawler` param you could instead check if the mob has the status effect.

AffectedArc07 replied:
Do this please

datlo commented:
I will eventually deconflict this and update to use a status effect, busy these days though

datlo commented:
Added an alert with a countdown to know exactly how long you have left in the vents. Refactored countdown alerts to be more flexile (was previously strictly for ghost poll notifications)
![image](https://user-images.githubusercontent.com/22121511/167922684-6f516954-72dc-4fab-8f49-224958331f2e.png)

datlo commented:
Ready for review/merge

farie82 commented on code/modules/atmospherics/machinery/components/unary_devices/unary_base.dm:
```diff
@@ -9,6 +9,8 @@
 
 	var/datum/pipeline/parent
 
+	var/welded = FALSE
```
I'd almost make you refactor the vent and scrubber to have a common parent. But I know the pain that would entail regarding map changes etc. I'm not sure this is the best approach though

farie82 commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -536,7 +544,15 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
 		client.eye = src
-
+	if(mind)
+		var/obj/effect/proc_holder/spell/reaper_lighter/S = locate(/obj/effect/proc_holder/spell/reaper_lighter) in mind.spell_list
```
I feel this should be handled with a signal instead. A signal that is being sent here that the mob exited a vent. And the effect, or spell, listening to this signal.

farie82 reviewed (changes requested):
The PR has a lot of unrelated changes. Try to keep it to a minimum for your next PR.
The effect does look interesting though. Good stuff

AffectedArc07 commented:
No activity in 7 days, closing.

Feel free to re-open in the future.
